changefeedccl: ShallowCopy RangeFeedCheckpoint before updating

stevendanna opened the pull request:
We suspect that the (*RangeFeedCheckpoint) used here may be shared by other clients and the rangefeed server side code when our rangefeed is being served by a local replica and thus the MuxRangeFeed RPC uses the local client optimization which avoid the network and also avoids marshling and unmarshaling the response across the RPC boundary.

This is currently our best theory for the source of the marshalling error seen in #159166. However, we have not yet been able to reproduce the crash locally.

Informs: https://github.com/cockroachdb/cockroach/issues/159166
Informs: https://github.com/cockroachdb/cockroach/issues/157882
Informs: https://github.com/cockroachdb/cockroach/issues/138895
Informs: https://github.com/cockroachdb/cockroach/issues/158166
Informs: https://github.com/cockroachlabs/support/issues/3506

Release note (bug fix): Fixes a bug which could lead to a node crash in the presence of a CHANGEFEED that uses the end_time option.

stevendanna committed:
changefeedccl: ShallowCopy RangeFeedCheckpoint before updating

We suspect that the (*RangeFeedCheckpoint) used here may be shared by other
clients and the rangefeed server side code when our rangefeed is being served by
a local replica and thus the MuxRangeFeed RPC uses the local client optimization
which avoid the network and also avoids marshling and unmarshaling the response
across the RPC boundary.

This is currently our best theory for the source of the marshalling error seen
in #159166:

1. A *RangeFeedCheckpoint is created by a processor and delivered to 2
  registrations. Both of those registrations are for spans that match the
  checkpoint span exactly and thus neither registration takes a shallow copy.

2. Registration 1 is for a local client and sends it to changefeed 1 via the
   local client optimization. As a result, changefeed 1's *RangeFeedCheckpoint
   is the same pointer that we have on the rangefeed server side.

3. Registration 2 begins to marshal the *RangeFeedCheckpoint and takes its Size.

4. Registration 1 overwrites (*RangeFeedCheckpoint).ResolvedTS to a value that
   has a larger encoded size.

5. Registration 2 continues its marshalling, now using the wrong size (and thus
   wrongly sized buffer). This leads to a panic.

However, we have not yet been able to reproduce the crash locally.

Informs: #159166
Informs: #157882
Informs: #138895
Informs: #158166
Informs: cockroachlabs/support#3506


Release note (bug fix): Fixes a bug which could lead to a node crash in the
presence of a CHANGEFEED that uses the end_time option.

Co-authored-by: Jeff Swenson <jeffswenson@betterthannull.com>

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei, @jeffswenson, and @stevendanna)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHCXGM1boagdCHa3q2:-OgHCXGM1boagdCHa3q3:bmd3xn5):*
The explanation you just typed out on slack would do well here too.
___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHBj7DEjWB9hJtkhJN:-OgHBj7DEjWB9hJtkhJO:b-u9xou8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe0cc24a358372432741550e056ab87b3f3fffa/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L786)):*
> ```Go
> 		//  - KV event: do nothing (we shouldn't emit this event)
> 		//  - Resolved event: advance this span to `boundary.ts` in the frontier
> 		checkCopyBoundary = func(e *kvevent.Event) (skipEvent, stopCopying bool, err error) {
> ```

why the pointer change?
___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHC623DKY8fOwgYouR:-OgHC6240cWfPuOUeWni:bo90ffg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe0cc24a358372432741550e056ab87b3f3fffa/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L786)):*
> ```Go
> 		//  - KV event: do nothing (we shouldn't emit this event)
> 		//  - Resolved event: advance this span to `boundary.ts` in the frontier
> 		checkCopyBoundary = func(e *kvevent.Event) (skipEvent, stopCopying bool, err error) {
> ```

one thought as a repro/regression test for this:

- make it a free-standing function
- exercise it in a unit test with concurrency (on the same event)
- data race before introducing `ResolvedBackward` proves that we did avoid a race here.
- could even throw in a goroutine marshaling the event in a tight loop to make sure the panic matches, but I wouldn't insist since a data race is already proof.
___
*[`pkg/ccl/changefeedccl/kvevent/event.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHCRwKCrrY1OE17PrL:-OgHCRwL1Foa1ii34ir9:b-fi7sdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe0cc24a358372432741550e056ab87b3f3fffa/pkg/ccl/changefeedccl/kvevent/event.go#L137)):*
> ```Go
> 
> // ResolvedBackwards moves the ResolvedTS of an underlying RangeFeedCheckpoint
> // event.
> ```

// Panics if the Event is not a Checkpoint.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei, @jeffswenson, and @tbg)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHCXGM1boagdCHa3q2:-OgHG4pH4gkW6t8l3EmM:b-896fix):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The explanation you just typed out on slack would do well here too.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/kvevent/event.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHCRwKCrrY1OE17PrL:-OgHG61O3f4UyzHwezGv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe0cc24a358372432741550e056ab87b3f3fffa/pkg/ccl/changefeedccl/kvevent/event.go#L137)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

// Panics if the Event is not a Checkpoint.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHBj7DEjWB9hJtkhJN:-OgHG6pM2gUgCK7xBbIZ:bkljjli) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe0cc24a358372432741550e056ab87b3f3fffa/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L786)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

why the pointer change?
</blockquote></details>

I think it is required for the e.ResolvedBackward call's assignment of `e.ev` to be effective?  (I think previous we were dealing with a copy of kvevent.Event that happen to point to the same `ev` and then doing an interior mutation of `ev`. Now we are overwriting `e.ev` so we need to be writing to the same `e`.
___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHC623DKY8fOwgYouR:-OgHGjQEApJCy1CSIk3G:bbo3zq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe0cc24a358372432741550e056ab87b3f3fffa/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L786)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

one thought as a repro/regression test for this:

- make it a free-standing function
- exercise it in a unit test with concurrency (on the same event)
- data race before introducing `ResolvedBackward` proves that we did avoid a race here.
- could even throw in a goroutine marshaling the event in a tight loop to make sure the panic matches, but I wouldn't insist since a data race is already proof.
</blockquote></details>

Oh nice idea.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159348)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei, @jeffswenson, and @stevendanna)

___
*[`-- commits` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-OgHGn-_1Xrnv1IITGb9:-OgHGn-_1Xrnv1IITGbA:b-r8pv98):*
I'm making this edit:

Informs: https://github.com/cockroachdb/cockroach/issues/159166
Informs: https://github.com/cockroachdb/cockroach/issues/157882
Informs: https://github.com/cockroachdb/cockroach/issues/138895
Informs: https://github.com/cockroachdb/cockroach/issues/158166
Informs: https://github.com/cockroachlabs/support/issues/3506


<!-- Sent from Reviewable.io -->

stevendanna committed:
changefeedccl: add regression test for event handling data race

This commit moves checkCopyBoundary to a freestanding function and adds a
regression test for the previous data race.

I've confirmed manually that this test does catch the previous race, but I've
left this as a commit ordered after the fix in case we want to _just_ ship a
custom build with the fix to an affected customer.

Epic: none
Release note: none

tbg reviewed (commented):
LGTM. I'll defer to @asg0451 for approval. Thanks Steven :heart: 

> @tbg reviewed 1 of 1 files at r3, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159348#-:-OgbnLLpBuv0XdOEjpUs:b-dg3j8p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei, @asg0451, and @jeffswenson)



<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM 🥳 

I think this change is safe. Long term I think we should rework the lifetimes, but agree this is the better change to backport.
